homeassistant/core.py

```diff
@@ -1207,51 +1208,25 @@ def _async_log_running_tasks(self, stage: str) -> None:
             _LOGGER.warning("Shutdown stage '%s': still running: %s", stage, task)
 
 
+@dataclasses.dataclass(frozen=True, kw_only=True)
 class Context:
     """The context that triggered something."""
 
-    def __init__(
-        self,
-        user_id: str | None = None,
-        parent_id: str | None = None,
-        id: str | None = None,  # pylint: disable=redefined-builtin
-    ) -> None:
-        """Init the context."""
-        self.id = id or ulid_now()
-        self.user_id = user_id
-        self.parent_id = parent_id
-        self.origin_event: Event[Any] | None = None
-
-    def __eq__(self, other: object) -> bool:
-        """Compare contexts."""
-        return isinstance(other, Context) and self.id == other.id
-
-    @cached_property
-    def _as_dict(self) -> dict[str, str | None]:
-        """Return a dictionary representation of the context.
-
-        Callers should be careful to not mutate the returned dictionary
-        as it will mutate the cached version.
-        """
-        return {
-            "id": self.id,
-            "parent_id": self.parent_id,
-            "user_id": self.user_id,
-        }
+    id: str = dataclasses.field(default_factory=ulid_now)
+    user_id: str | None = dataclasses.field(default=None, compare=False)
+    parent_id: str | None = dataclasses.field(default=None, compare=False)
+    origin_event: Event[Any] | None = dataclasses.field(default=None, compare=False)
 
     def as_dict(self) -> ReadOnlyDict[str, str | None]:
         """Return a ReadOnlyDict representation of the context."""
-        return self._as_read_only_dict
-
-    @cached_property
-    def _as_read_only_dict(self) -> ReadOnlyDict[str, str | None]:
-        """Return a ReadOnlyDict representation of the context."""
-        return ReadOnlyDict(self._as_dict)
+        dict_ = dataclasses.asdict(self)
```

Comment:
These objects are performance sensitive since we can create 10000s of them per minute.

`dataclasses.asdict` have some known performance problems, and while its better after https://github.com/python/cpython/issues/114087 its probably better to write it out for these object since they are used everywhere